src/ast.rs

```diff
@@ -436,6 +439,35 @@ impl<'prev, 'subs> ArgScope<'prev, 'subs> for Option<ArgScopeStack<'prev, 'subs>
     }
 }
 
+#[derive(Debug, Copy, Clone)]
+struct DemangleState {
+    /// How deep in the demangling are we?
+    pub recursion_level: u32,
+}
+
+/// An RAII type to automatically check the recursion level against the
+/// maximum. If the maximum has been crossed, return an error. Otherwise,
+/// increment the level upon construction, and decrement it upon destruction.
+struct AutoParseDemangle(Rc<Cell<DemangleState>>);
```

Comment:
I think we should be able to avoid the `Rc` by making `AutoParseDemangle` wrap the `DemangleContext` and implement deref[mut] to it. Then we use it, instead of the original `ctx` everywhere.

```rust
fn foo(ctx: &DemangleContext<'a, W>) -> Result<...> {
    let ctx = begin_demangle!(ctx, ...);
    // ctx is actually an `AutoParseDemangle`, but we can pretty much
    // use it everywhere we would pass a `DemangleContext`, and its
    // destructor will decrement the recursion level when we return.
}
```

Finally, let's rename `AutoParseDemangle` to `AutoDemangleRecursion` or something like that, since it exists to guard against recursion in the demangling phase, and doesn't really have to do with parsing.